File: Code/python/dataloader.py
import sys
from pathlib import Path

build_path = Path(__file__).resolve().parents[1] / "build" / "bindings"
sys.path.append(str(build_path))

import tinytorch as tt

print("=" * 60)
print("DataLoader & ImageDataset Demo")
print("=" * 60)

# 1. ImageDataset with Mock Data and Transforms
print("Initializing ImageDataset with RandomCrop...")
image_paths = ["img1.jpg", "img2.jpg", "img3.jpg", "img4.jpg", "img5.jpg"]
labels = [tt.Tensor([0], [1]), tt.Tensor([1], [1]), tt.Tensor([0], [1]), tt.Tensor([1], [1]), tt.Tensor([2], [1])]

transform = tt.RandomCrop((28, 28), padding=4, seed=42)
img_dataset = tt.ImageDataset(image_paths, labels, transform)

print(f"Dataset contains {len(img_dataset)} samples")

# 2. DataLoader (Batch Size = 2, Shuffle = True)
print("\nInitializing DataLoader (batch_size=2, shuffle=True)...")
loader = tt.DataLoader(img_dataset, batch_size=2, shuffle=True, seed=99)

print(f"Total Batches per epoch: {len(loader)}")

# 3. Simulate Training Loop (Lazy Evaluation)
print("\n--- Starting Training Epoch ---")
for batch_idx, batch in enumerate(loader):
    features, targets = batch[0], batch[1]
    
    print(f"\nBatch {batch_idx}:")
    print(f"  Features Shape: {features.shape}  <-- Note the Batch Dimension!")
    print(f"  Targets Shape : {targets.shape}")
    print(f"  Targets Values: {targets.tolist()}")

print("\nTraining Epoch Finished!")

File: Code/bindings/data/dataloader.cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#include "data/dataloader.h"
#include "data/image_dataset.h"

namespace py = pybind11;

void bind_dataloader(py::module_& m) {
    py::class_<ImageDataset, Dataset, std::shared_ptr<ImageDataset>>(m, "ImageDataset")
        .def(py::init<const std::vector<std::string>&, const std::vector<Tensor>&,
                      std::shared_ptr<Transform>>(),
             py::arg("image_paths"), py::arg("labels"), py::arg("transform") = nullptr)
        .def("__len__", &ImageDataset::size)
        .def("__getitem__", &ImageDataset::get_item);

    py::class_<DataLoader, std::shared_ptr<DataLoader>>(m, "DataLoader")
        .def(py::init<const Dataset&, Tensor::size_type, bool, std::optional<unsigned int>>(),
             py::arg("dataset"), py::arg("batch_size"), py::arg("shuffle") = false,
             py::arg("seed") = py::none())
        .def("__len__", &DataLoader::size)
        .def(
            "__iter__",
            [](const DataLoader& loader) {
                return py::make_iterator(loader.begin(), loader.end());
            },
            py::keep_alive<0, 1>());
}
